File: Chapter 02/TemplateMethodDesignPatternExample.py
class Sport:
    def initialize(self):
        pass
    def start_play(self):
        pass
    def end_play(self):
        pass
    def play(self):
        self.initialize()
        self.start_play()
        self.end_play()

class Basketball(Sport):
    def initialize(self):
        print("Basketball court is ready! Start playing.")
    def start_play(self):
        print("Basketball Game Started. Enjoy the game!")
    def end_play(self):
        print("Basketball Game Finished!")

class Pickleball(Sport):
    def initialize(self):
        print("Pickleball court is ready! Start playing.")
    def start_play(self):
        print("Pickleball Game Started. Enjoy the game!")
    def end_play(self):
        print("Pickleball Game Finished!")

if __name__ == "__main__":
    basketball_game = Basketball()
    pickleball_game = Pickleball()

    print("Playing Basketball...")
    basketball_game.play()

    print("\nPlaying Pickleball...")
    pickleball_game.play()

File: Chapter 02/TemplateMethodDesignPatternExample.java
abstract class Sport {
    abstract void initialize();
    abstract void startPlay();
    abstract void endPlay();

    public final void play() {
        initialize();
        startPlay();
        endPlay();
    }
}


class Basketball extends Sport {
    void initialize() {
        System.out.println("Basketball court is ready! Start playing.");
    }

    void startPlay() {
        System.out.println("Basketball game started. Enjoy the game!");
    }

    void endPlay() {
        System.out.println("Basketball game finished!");
    }
}

class Pickleball extends Sport {
    void initialize() {
        System.out.println("Pickleball court is ready! Start playing.");
    }

    void startPlay() {
        System.out.println("Pickleball Game Started. Enjoy the game!");
    }

    void endPlay() {
        System.out.println("Pickleball Game Finished!");
    }
}


public class TemplateMethodDesignPatternExample {
    public static void main(String[] args) {
        Sport basketballGame = new Basketball();
        Sport pickleballGame = new Pickleball();

        System.out.println("Playing Basketball...");
        basketballGame.play();

        System.out.println("\nPlaying Pickleball...");
        pickleballGame.play();
    }
}

